op-supervisor/supervisor/backend/cross/cycle_test.go

```diff
@@ -0,0 +1,511 @@
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 0),
+					},
+				},
+			},
+			msg: "expected no cycle found for one basic and one exec log",
+		},
+		{
+			name: "first log is exec",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 1,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("2", 0),
+					},
+				},
+				"2": {
+					logCount: 1,
+					messages: nil,
+				},
+			},
+			msg: "expected no cycle found first log is exec",
+		},
+		{
+			name: "cycle through different timestamp",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("2", 0),
+						1: execMsgWithTimestamp("2", 1, 99),
+					},
+				},
+				"2": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("1", 1),
+					},
+				},
+			},
+			msg: "expected no cycle detection error for cycle through messages with different timestamps",
+		},
+	}
+	runHazardCycleChecksTestCaseGroup(t, "NoCycle", tests)
+}
+
+func TestHazardCycleChecksCycle(t *testing.T) {
+	tests := []hazardCycleChecksTestCase{
+		{
+			name: "2-cycle in single chain with first log",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 3,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("1", 2),
+						2: execMsg("1", 0),
+					},
+				},
+			},
+			expectErr: ErrCycle,
+			msg:       "expected cycle detection error",
+		},
+		{
+			name: "2-cycle in single chain with first log, adjacent",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						0: execMsg("1", 1),
+						1: execMsg("1", 0),
+					},
+				},
+			},
+			expectErr: ErrCycle,
+			msg:       "expected cycle detection error",
+		},
+		{
+			name: "2-cycle in single chain, not first, adjacent",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 3,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 2),
+						2: execMsg("1", 1),
+					},
+				},
+			},
+			expectErr: ErrCycle,
+			msg:       "expected cycle detection error",
+		},
+		{
+			name: "2-cycle in single chain, not first, not adjacent",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 4,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 3),
+						3: execMsg("1", 1),
+					},
+				},
+			},
+			expectErr: ErrCycle,
+			msg:       "expected cycle detection error",
+		},
+		{
+			name: "2-cycle across chains",
+			chainBlocks: map[string]chainBlockDef{
+				"1": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("2", 1),
+					},
+				},
+				"2": {
+					logCount: 2,
+					messages: map[uint32]*types.ExecutingMessage{
+						1: execMsg("1", 1),
+					},
+				},
```

Comment:
Maybe it makes more sense if the `key == 0`, to create a cycle? Otherwise the cycle collapses to the same kind of self-referential log-index case.
So that would be `1: execMsg("1", 1),`